torch/csrc/jit/codegen/cuda/lower_unroll.cpp

```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+
+    const auto new_loop = ir_builder.create<kir::ForLoop>(
+        fl->index(),
+        fl->iter_domain(),
+        vectorize && has_vectorize_op,
+        extent,
+        false,
+        shift);
+
+    for (auto expr : fl->body().exprs()) {
+      new_loop->body().push_back(expr);
+    }
+
+    parent_ite->thenBody().push_back(new_loop);
+  }
+}
+
+// Find any child For-Loops
+// Add remaining expressions to new parent For-Loop
+std::vector<kir::ForLoop*> parseVectorizedForLoop(
+    const kir::ForLoop* for_loop,
+    kir::ForLoop* new_loop) {
+  std::vector<kir::ForLoop*> loops;
+  for (auto expr : for_loop->body().exprs()) {
+    if (auto nested_for_loop = dynamic_cast<kir::ForLoop*>(expr)) {
+      loops.push_back(nested_for_loop);
+    } else {
+      new_loop->body().push_back(expr);
+    }
+  }
+  return loops;
+}
+
+// Find the first vectorize set - either read or write
+// Add child For-Loop to loop_structure
+// Enable vectorize flag in child For-Loop
+kir::Expr* findVectorizedSet(
+    std::vector<kir::ForLoop*>& loop_structure,
+    const std::vector<kir::ForLoop*>& for_loops) {
+  for (auto fl : for_loops) {
+    auto first_expr = fl->body().exprs().front();
+    bool has_vectorize_op =
+        (first_expr->isA<kir::UnaryOp>() &&
+         first_expr->as<kir::UnaryOp>()->operation() == UnaryOpType::Set &&
+         fl->iter_domain()->parallelType() ==
+             ParallelType::MisalignedVectorize);
+    if (has_vectorize_op) {
+      loop_structure.push_back(fl);
+      return first_expr;
+    }
+  }
+  return nullptr;
+}
+
+// Get full extent for the inner-most, merged root domain
+kir::Val* getVectorizeExtent(
+    kir::TensorView* tv,
+    const std::vector<kir::Val*>& indices) {
+  auto domain = tv->domain()->hasRFactor() ? tv->domain()->rfactorDomain()
+                                           : tv->domain()->rootDomain();
+
+  TORCH_INTERNAL_ASSERT(domain.size() == indices.size());
+
+  bool is_contiguous = true;
+  for (auto status : tv->domain()->contiguity()) {
+    is_contiguous &= status;
+  }
+
+  // If the tensorview is not contiguous, return inner-most root domain extent
+  if (!is_contiguous) {
+    return domain.back()->extent();
+  }
+
+  const auto gpu_lower = GpuLower::current();
+  kir::IrBuilder ir_builder(gpu_lower->kernel());
+
+  // Calculate extent of merged root domains
+  kir::Val* extent = nullptr;
+  for (int i = int(domain.size()) - 1; i >= 0; --i) {
+    auto root_id = domain.at(i);
+    if (root_id->isBroadcast() || root_id->isReduction() ||
+        gpu_lower->trivialReductionInfo().isDerived(root_id)) {
+      continue;
+    } else if (extent == nullptr) {
+      extent = root_id->extent();
+    } else if (extent != nullptr && indices.at(i)->isZeroInt()) {
+      // This root id must be merged and contiguous. Expand the
+      // vectorization partition.
+      extent = ir_builder.mulExpr(extent, root_id->extent());
+    } else {
+      break;
```

Comment:
Unclear to me why we'd hit break from reading this code. Add comments.

Reply:
@naoyam why is there a break here?

Reply:
The loop here finds the maximum consecutive root axes that are contiguous. For global indexing, `indices.at(i)` being zero means it's contiguous, so that's what I used here. However, I don't understand why I did that rather than just using the contiguity vector. I'll clean up this part.

Reply:
See #826